store.rb

```diff
@@ -128,4 +181,43 @@ def get_node(node)
       raise e
     end
   end
+
+  # immediately update cache if node joins/leaves
+  def watch()
+    return if @zk.nil?
+    @cache_root_watcher ||= @zk.register("/", :only => :child) do |event|
+      @log.info "Children added/deleted"
+      reload_instances
+    end
+  end
+
+  def reload_instances()
+    # Here we use local time to preven race condition
+    # Basically cache fetch thread or zookeeper watch callback
+    # both will call this to refresh cache. Depending on which
+    # finishes first our cache will get set by the slower one.
+    # So in order to prevent setting cache to an older result,
+    # we set both cache and the timestamp of that version fetched
+    # Since timestamp will be monotonically increasing, we are
+    # sure that cache set will always have newer versions
+
+    fetch_start_time = Time.now
+    instances = load_instances_from_zk()
+    @cache_mutex.synchronize do
+      if fetch_start_time > @cache_results_last_fetched_time then
+        @cache_results_last_fetched_time = fetch_start_time
+        @cache_results = instances
+      end
+    end
+  end
+
+  def start_fetch_thread()
+    @cache_fetch_thread = Thread.new do
+      while @cache_fetch_thread_should_run do
```

Comment:
you should probably wrap this entire loop in an exception handler. i'm not sure what to do about the exceptions, but at last logging them might be nice